core/src/main/scala/org/apache/spark/executor/BaseCoarseGrainedExecutorBackend.scala

```diff
@@ -0,0 +1,232 @@
+        executor.launchTask(this, taskDesc)
+      }
+
+    case KillTask(taskId, _, interruptThread, reason) =>
+      if (executor == null) {
+        exitExecutor(1, "Received KillTask command but executor was null")
+      } else {
+        executor.killTask(taskId, interruptThread, reason)
+      }
+
+    case StopExecutor =>
+      stopping.set(true)
+      logInfo("Driver commanded a shutdown")
+      // Cannot shutdown here because an ack may need to be sent back to the caller. So send
+      // a message to self to actually do the shutdown.
+      self.send(Shutdown)
+
+    case Shutdown =>
+      stopping.set(true)
+      new Thread("CoarseGrainedExecutorBackend-stop-executor") {
+        override def run(): Unit = {
+          // executor.stop() will call `SparkEnv.stop()` which waits until RpcEnv stops totally.
+          // However, if `executor.stop()` runs in some thread of RpcEnv, RpcEnv won't be able to
+          // stop until `executor.stop()` returns, which becomes a dead-lock (See SPARK-14180).
+          // Therefore, we put this line in a new thread.
+          executor.stop()
+        }
+      }.start()
+
+    case UpdateDelegationTokens(tokenBytes) =>
+      logInfo(s"Received tokens of ${tokenBytes.length} bytes")
+      SparkHadoopUtil.get.addDelegationTokens(tokenBytes, env.conf)
+  }
+
+  override def onDisconnected(remoteAddress: RpcAddress): Unit = {
+    if (stopping.get()) {
+      logInfo(s"Driver from $remoteAddress disconnected during shutdown")
+    } else if (driver.exists(_.address == remoteAddress)) {
+      exitExecutor(1, s"Driver $remoteAddress disassociated! Shutting down.", null,
+        notifyDriver = false)
+    } else {
+      logWarning(s"An unknown ($remoteAddress) driver disconnected.")
+    }
+  }
+
+  override def statusUpdate(taskId: Long, state: TaskState, data: ByteBuffer) {
+    val msg = StatusUpdate(executorId, taskId, state, data)
+    driver match {
+      case Some(driverRef) => driverRef.send(msg)
+      case None => logWarning(s"Drop $msg because has not yet connected to driver")
+    }
+  }
+
+  def extractLogUrls: Map[String, String]
+
+  def extractAttributes: Map[String, String]
+
+  /**
+   * This function can be overloaded by other child classes to handle
+   * executor exits differently. For e.g. when an executor goes down,
+   * back-end may not want to take the parent process down.
+   */
+  protected def exitExecutor(
+      code: Int,
+      reason: String,
+      throwable: Throwable = null,
+      notifyDriver: Boolean = true) = {
+    val message = "Executor self-exiting due to : " + reason
+    if (throwable != null) {
+      logError(message, throwable)
+    } else {
+      logError(message)
+    }
+
+    if (notifyDriver && driver.nonEmpty) {
+      driver.get.send(RemoveExecutor(executorId, new ExecutorLossReason(reason)))
+    }
+
+    System.exit(code)
+  }
+}
+
+object BaseCoarseGrainedExecutorBackend {
+  private[spark] def run(
```

Comment:
Do you actually need different command line parsing for the YARN version? Up to now they've been the same, so it seems to me they should remain the same.

So if instead of this, you add `main(args, backendCreateFn)` to `CoarseGrainedExecutorBackend`, you could share more code.